pygsti/forwardsims/torchfwdsim.py

```diff
@@ -0,0 +1,237 @@
+            spc = list(expanded_circuit_outcomes.keys())[0]
+            c = StatelessCircuit(spc)
+            circuits.append(c)
+        self.circuits = circuits
+
+        self.param_metadata = []
+        for lbl, obj in model._iter_parameterized_objs():
+            assert isinstance(obj, Torchable)
+            param_type = type(obj)
+            param_data = (lbl, param_type) + (obj.stateless_data(),)
+            self.param_metadata.append(param_data)
+        self.num_params = len(self.param_metadata)
+        return
+    
+    def get_free_parameters(self, model: ExplicitOpModel):
+        """
+        Return an ordered dict that maps pyGSTi Labels to PyTorch Tensors.
+        The Labels correspond to parameterized objects in "model".
+        The Tensors correspond to the current values of an object's parameters.
+        For the purposes of forward simulation, we intend that the following 
+        equivalence holds:
+        
+            model == (self, [dict returned by this function]).
+            
+        That said, the values in this function's returned dict need to be
+        formatted by get_torch_cache BEFORE being used in forward simulation.
+        """
+        free_params = OrderedDict()
+        for i, (lbl, obj) in enumerate(model._iter_parameterized_objs()):
+            gpind = obj.gpindices_as_array()
+            vec = obj.to_vector()
+            vec = torch.from_numpy(vec)
+            assert int(gpind.size) == int(np.prod(vec.shape))
+            # ^ a sanity check that we're interpreting the results of obj.to_vector()
+            #   correctly. Future implementations might need us to also keep track of
+            #   the "gpind" variable. Right now we get around NOT using that variable
+            #   by using an OrderedDict and by iterating over parameterized objects in
+            #   the same way that "model"s does.
+            assert self.param_metadata[i][0] == lbl
+            # ^ If this check fails then it invalidates our assumptions about how
+            #   we're using OrderedDict objects.
+            free_params[lbl] = vec
+        return free_params
+
+    def get_torch_cache(self, free_params: OrderedDict[Label, torch.Tensor], grad: bool):
+        """
+        Returns a dict mapping pyGSTi Labels to PyTorch tensors. The dict makes it easy
+        to simulate a stateful model implied by (self, free_params). It is obtained by
+        applying invertible transformations --- defined in various ModelMember subclasses
+        --- on the tensors stored in free_params.
+
+        If ``grad`` is True, then the values in the returned dict are preparred for use
+        in PyTorch's backpropogation functionality. If we want to compute a Jacobian of
+        circuit outcome probabilities then such functionality is actually NOT needed.
+        Therefore for purposes of computing Jacobians this should be set to False.
+        """
```

Comment:
In general I think the answer is yes. Looking at a single circuit, the jacobian itself will have dimensions that are $d \times N_p$, where d is the Hilbert space dimension, and $d=2^n$ where $n$ is the number of qubits, and $N_p$ is the number of model parameters. The number of model parameters grows much faster than the dimension of the Hilbert space ($\mathcal(O)(16^n)$ vs $\mathcal(O)(2^n)$), so this will generically be much wider than it is tall. Of course, that is assuming fully parameterized models. Reduced models (like weight 1+2 error generator models) scale differently. There is also the confounding factor that we only need to compute the cost function for outcomes we actually observe, in the context of MLE. Anyhow, all of this is to say that while I suspect we will find reverse-mode advantageous, figuring out exactly the parameter regimes where that is true will indeed require experimentation. 

Thoughts on adding a handle on this behavior?

Reply:
@coreyostrove, the Jacobian calculation isn't done one circuit at a time; it's done with consideration to all circuits at once. I agree that if we're looking at an individual circuit then the map from model parameters to circuit outcome probabilities would be wide.

That said, I've modified the code to automatically select between forward-mode AD and reverse-mode AD depending on the aspect ratio of the Jacobian :)

Reply:
Awesome, great work!